src/HexagonOptimize.cpp

```diff
@@ -1605,6 +1618,320 @@ class OptimizeShuffles : public IRMutator {
+                            continue;
+                        }
+                        for (size_t k = 0; k < mpy_size; k++) {
+                            Expr load0 = loads[k];
+                            if (!load0.defined() || !cmp_basediff_const(load1, load0, load0.type().bits()/8)) {
+                                continue;
+                            }
+                            Expr new_expr = native_interleave(Call::make(op->type, "halide.hexagon.vtmpy" + vtmpy_suffix, {mpys[k].first, mpys[i].first, mpys[k].second, mpys[j].second}, Call::PureExtern));
+                            Expr sum;
+                            for (size_t l = 0; l < mpy_size; l++) {
+                                if (l==i || l==j || l==k)
+                                    continue;
+                                if (sum.defined()) {
+                                    sum = sum + lossless_cast(op->type, mpys[l].first)*lossless_cast(op->type, mpys[l].second);
+                                } else {
+                                    sum = lossless_cast(op->type, mpys[l].first)*lossless_cast(op->type, mpys[l].second);
+                                }
+                            }
+                            if (sum.defined()) {
+                                new_expr = new_expr + sum;
+                            }
+                            if (rest.defined()) {
+                                new_expr = new_expr + rest;
+                            }
+                            expr = mutate(new_expr);
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+
+        IRMutator::visit(op);
+    }
+};
+
+// Convert some expressions to an equivalent form which could get better
+// optimized in later stages for hexagon
+class RearrangeExpressions : public IRMutator {
+private:
+    using IRMutator::visit;
+
+    void visit(const Mul *op) {
+        IRMutator::visit(op);
+
+        if (op->type.is_vector()) {
+            op = expr.as<Mul>();
+            vector<Expr> matches;
+            if (op->a.as<Broadcast>()) {
+                // Ensures broadcasts always occurs as op1 not op0
+                expr = mutate(op->b * op->a);
+                return;
+            }
+            // Distributing broadcasts helps creating more vmpa because of more adds of muls
+            if (op->b.as<Broadcast>()) {
+                static const vector<Expr> patterns = {
+                    (wild_i8x + wild_i8x),
+                    (wild_i16x + wild_i16x),
+                    (wild_u8x + wild_u8x),
+                    (wild_u16x + wild_u16x),
+                };
+                for (const Expr &p : patterns) {
+                    if(expr_match(p, op->a, matches)) {
+                        expr = mutate(simplify(matches[0] * op->b) + simplify(matches[1] * op->b));
```

Comment:
in case match[0] is of the form -> bc(int)*vector then simplify will combine the two bc into one